### PR TITLE
Fetch address from PCIU service

### PR DESCRIPTION
**Pull request by raquelromano-gov**
Connects to https://github.com/department-of-veterans-affairs/vets.gov-team/issues/4680

Instead of getting the letter address from the letters `destination` property, fetch it from the PCIU mailingAddress endpoint so that the format matches what's expected by the address update endpoint.

Not done in this PR: the loading indicator should now rely on the address availability instead of the letters availability.

Still needs unit tests.

### CODE REVIEW

File: src/js/letters/utils/constants.js [hunk 1]
```diff
@@ -4,10 +4,76 @@ export const BACKEND_SERVICE_ERROR = 'BACKEND_SERVICE_ERROR'; // 503 or 504
...
+  WI: 'Wisconsin',
+  WY: 'Wyoming',
+  AS: 'American Samoa',
+  DC: 'District of Columbia',
+  FM: 'Federated States of Micronesia',
+  GU: 'Guam',
+  MH: 'Marshall Islands',
+  MP: 'Northern Mariana Islands',
+  PW: 'Palau',
+  PR: '',
```
Comment (annekainicUSDS) [inline src/js/letters/utils/constants.js]: @raquelromano-gov do you still have the email with the rest of the translations for these states? Thanks!

File: src/js/letters/containers/AddressSection.jsx [hunk 2]
```diff
@@ -53,7 +81,7 @@ export class AddressSection extends React.Component {
             Downloaded documents will list your address as:
           </p>
           <div className="address-block">
-            <h5 className="letters-address">{(destination.fullName || '').toLowerCase()}</h5>
+            <h5 className="letters-address">{(this.props.fullName || '').toLowerCase()}</h5>
```
Comment (cvalarida) [inline src/js/letters/containers/AddressSection.jsx]: So this took me a long time to figure out what `fullName` belonged to. Can we change this to `recipientName` or something to make it clear that it sort of "belongs" in the address? Or does that not make as much sense?

File: src/js/letters/utils/helpers.jsx [hunk 3]
```diff
@@ -244,3 +245,30 @@ export const benefitOptionsMap = {
...
+  }
+  const parts = [
+    address.zipCode,
+    address.zipSuffix ? `-${address.zipSuffix}` : ''
+  ];
+  return parts.join('');
+}
+
+export function getStateName(stateCode) {
+  return STATE_CODE_TO_NAME[stateCode] || '';
```
Comment (cvalarida) [inline src/js/letters/utils/helpers.jsx]: Should we throw something to sentry if we're trying to get a state name that isn't in our list? That might be better in another PR, but...just a thought. :man_shrugging:

**Review by cvalarida — APPROVED** [review]
Phew! Sorry it took so long! I had to do a lot of digging.

I had a bunch of questions and a couple nits, but I don't think they're worth stopping the merge, so I'm giving this the 👍

File: src/js/letters/actions/letters.js [hunk 4]
```diff
@@ -45,14 +48,60 @@ export function getLetterList() {
...
         }
         return Promise.reject(
-          new Error(`vets_letters_error_server_get: ${error.status}`)
+          new Error('vets_letters_error_server_get')
         );
       })
       .catch((error) => {
-        Raven.captureException(error);
-        return dispatch({ type: GET_LETTERS_FAILURE });
+        if (error.message.match('vets_letters_error_server_get')) {
```
[the remark this answers is not in the document]
Comment (raquelromano-gov) [reply src/js/letters/actions/letters.js]: Yeah...I  put these in here, and felt weird about it, but saw that we're using these elsewhere in the code to make it easy to search sentry for particular errors. It didn't bother me enough to not write it :)

File: src/js/letters/components/Address.jsx [hunk 5]
```diff
@@ -47,18 +47,10 @@ class Address extends React.Component {
...
 
-    if (this.props.required || !isBlankAddress(this.props.value)) {
-      isValid = isValid && isNotBlank(postalCodeField);
-    }
-
     if (this.props.value.country === 'USA' && isNotBlank(postalCodeField)) {
       isValid = isValid && isValidUSZipCode(postalCodeField);
     }
 
-    if (this.props.value.country === 'CAN' && isNotBlank(postalCodeField)) {
```
[the remark this answers is not in the document]
Comment (raquelromano-gov) [reply src/js/letters/components/Address.jsx]: I think I started removing this when I realized there are no international postal codes allowed. I believe that code was just lifted from the common `Address.jsx` and was trying to get rid of some of the bits we don't need in this custom component.

File: src/js/letters/components/Address.jsx [hunk 6]
```diff
@@ -69,75 +61,104 @@ class Address extends React.Component {
...
+    const addressType = this.props.value.type;
+
+    // Depending on the type of address, the fields for state and city
+    // are called different things. Set new variables for the name of
+    // the appropriate city and state field, and use those below for
+    // the field components.
+    let cityField;
+    let stateField;
+    if (addressType === 'MILITARY') {
+      cityField = 'militaryPostOfficeTypeCode';
```
[the remark this answers is not in the document]
Comment (raquelromano-gov) [reply src/js/letters/components/Address.jsx]: Hm...now that you point this out, it's wrong. We got clarity on requirements since this was written: https://github.com/department-of-veterans-affairs/vets.gov-team/issues/4110#issuecomment-328640426

File: src/js/letters/components/Address.jsx [hunk 7]
```diff
@@ -69,75 +61,104 @@ class Address extends React.Component {
   }
 
   render() {
+    const addressType = this.props.value.type;
```
[the remark this answers is not in the document]
Comment (raquelromano-gov) [reply src/js/letters/components/Address.jsx]: Yes and no. We do get it from the API, but it should be updated in the background when the user updates fields since we infer it based on the user data. The gory details are in https://github.com/department-of-veterans-affairs/vets.gov-team/issues/4110#issuecomment-328640426

File: src/js/letters/components/Address.jsx [hunk 8]
```diff
@@ -69,75 +61,104 @@ class Address extends React.Component {
...
+      cityField = 'militaryPostOfficeTypeCode';
+      stateField = 'militaryStateCode';
+    } else {
+      cityField = 'city';
+      stateField = 'state';
+    }
+
+    // Our hard-coded list of countries has the value for the U.S. as
+    // 'USA', but the value sent to us from EVSS is 'US'. This will cause
+    // the field to not populate correctly.
```
[the remark this answers is not in the document]
Comment (raquelromano-gov) [reply src/js/letters/components/Address.jsx]: I think this comment can go away.  The country name list is fetched from a backend service, and in that service the value is "USA" 
But confusingly, in some test data we have in PINT, there was a value of US stored, but that was when we were getting the address from the letter list endpoint, which we're no longer doing.

So bottom line: it should not be hard-coded anywhere, just fetched, and it should always be "USA"

File: src/js/letters/components/Address.jsx [hunk 9]
```diff
@@ -69,75 +61,104 @@ class Address extends React.Component {
...
         stateList = stateList.filter(state => state.value === 'AE' || state.value === 'AP' || state.value === 'AA');
       }
     }
 
-    const stateProvince = _.hasIn(states, this.props.value.country === 'US' ? 'USA' : this.props.value.country)
-      ? (<ErrorableSelect errorMessage={this.isValidAddressField(this.props.value.state) ? undefined : 'Please enter a valid state/province'}
-        label={this.props.value.country === 'CAN' ? 'Province' : 'State'}
+    const stateProvince = _.hasIn(states, selectedCountry)
+      ? (<ErrorableSelect errorMessage={this.isValidAddressField(this.props.value[stateField]) ? undefined : 'Please enter a valid state/province'}
+        label="State"
```
[the remark this answers is not in the document]
Comment (raquelromano-gov) [reply src/js/letters/components/Address.jsx]: This can also go away (see same comment in the issue above): there shall be no province or state names for international addresses because the VA format says so.

File: src/js/letters/containers/AddressSection.jsx (hunk 2 above)
Comment (raquelromano-gov) [reply src/js/letters/containers/AddressSection.jsx]: Sure--it comes to use in the response as `fullName` but sometimes these weirdly-named response properties are worth renaming in our props.

File: src/js/letters/utils/constants.js (hunk 1 above)
Comment (raquelromano-gov) [reply src/js/letters/utils/constants.js]: Should be here https://github.com/department-of-veterans-affairs/vets.gov-team/issues/4110#issuecomment-329359875 but yes, I can make sure you are all on that thread (later tonight)

File: src/js/letters/utils/helpers.jsx (hunk 3 above)
Comment (raquelromano-gov) [reply src/js/letters/utils/helpers.jsx]: 👍  SGTM either now or eventually, but maybe now is best

File: src/js/letters/utils/constants.js (hunk 1 above)
Comment (annekainicUSDS) [reply src/js/letters/utils/constants.js]: Perfect, thanks!

File: src/js/letters/actions/letters.js [hunk 10]
```diff
@@ -7,6 +7,8 @@ import {
   INVALID_ADDRESS_PROPERTY,
   GET_LETTERS_FAILURE,
   GET_LETTERS_SUCCESS,
+  GET_ADDRESS_FAILURE,
+  GET_ADDRESS_SUCCESS,
```
[the remark this answers is not in the document]
Comment (annekainicUSDS) [reply src/js/letters/actions/letters.js]: Ah interesting, I didn't think of reading it that way. I see what you mean, I'm open to creating a ticket to address this in a separate PR!

File: src/js/letters/actions/letters.js [hunk 11]
```diff
@@ -36,6 +38,7 @@ export function getLetterList() {
             // Backend authentication problem
             return dispatch({ type: BACKEND_AUTHENTICATION_ERROR });
           }
+          // TODO: remove this since we no longer use the letters destination properties
```
[the remark this answers is not in the document]
Comment (annekainicUSDS) [reply src/js/letters/actions/letters.js]: Ah yes, good catch. I will remove this one.

File: src/js/letters/actions/letters.js [hunk 12]
```diff
@@ -45,14 +48,60 @@ export function getLetterList() {
...
+          return Promise.reject(
+            new Error(`vets_letters_error_server_get: ${error.status}`)
+          );
         }
         return Promise.reject(
-          new Error(`vets_letters_error_server_get: ${error.status}`)
+          new Error('vets_letters_error_server_get')
         );
       })
       .catch((error) => {
```
[the remark this answers is not in the document]
Comment (annekainicUSDS) [reply src/js/letters/actions/letters.js]: Hehe yep, I can change that indentation

File: src/js/letters/components/Address.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (annekainicUSDS) [reply src/js/letters/components/Address.jsx]: Exactly, what Raquel said here. We started this new component by just copying the common Address component, which has additional things we don't need to keep in here. While it's not directly related to the purpose of this PR, it's just cleaning up a line we don't need.

File: src/js/letters/components/Address.jsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (annekainicUSDS) [reply src/js/letters/components/Address.jsx]: Yep, that work hasn't been done yet and isn't in this PR. I assume we will update this value by dispatching an action for this, so it won't be changed within this component anyway; this component will only ever receive the value for the type of address as a prop.

File: src/js/letters/components/Address.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (annekainicUSDS) [reply src/js/letters/components/Address.jsx]: Hm sorry missing what's wrong here @raquelromano-gov? I see `militaryStateCode` corresponds with the `state` field, so it would seem `militaryPostOfficeTypeCode` corresponds to the city field (at least this is what we have done on all other forms, where the options of "APO", "FPO", and "DPO" stand in for the city). The logic is a bit odd, how `militaryStateCode` is asked first. This would change the order of our inputs, and change the conditional logic of when inputs are displayed or hidden. This makes things more complicated without knowing the type of address first. Might be worth talking this through with @goldenmeanie tomorrow.

File: src/js/letters/components/Address.jsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (annekainicUSDS) [reply src/js/letters/components/Address.jsx]: Yes, this craziness goes away once this PR gets merged: https://github.com/department-of-veterans-affairs/vets-website/pull/6311/files (see comment one line below on line 83 :) ).

File: src/js/letters/actions/letters.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (annekainicUSDS) [reply src/js/letters/actions/letters.js]: So apparently npm lint doesn't like that other indentation, so I think I have to keep it the way it is here.

File: src/js/letters/containers/AddressSection.jsx (hunk 2 above)
Comment (annekainicUSDS) [reply src/js/letters/containers/AddressSection.jsx]: Yep, can change.

File: src/js/letters/containers/AddressSection.jsx [hunk 13]
```diff
@@ -13,34 +20,55 @@ export class AddressSection extends React.Component {
   }
```
Comment (markgreenburg) [inline src/js/letters/containers/AddressSection.jsx]: Not sure if it would make more sense for clarity to address in a separate tidy-up PR but since you're here...the state's getting initialized with `{ isEditAddressing: false }` - while I like this version better I think you meant `{ isEditingAddress: false }` 😛

Comment (annekainicUSDS) [reply src/js/letters/containers/AddressSection.jsx]: Ah yes! good catch, will fix this

**Comment by annekainicUSDS** [comment]
Going to merge this PR for now because it is blocking other work. Should have made all the fixes identified here in code review except for the `militaryPostOfficeTypeCode` question @raquelromano-gov brought up. Not sure what the right fix is for that at the moment, but can make that in a separate PR once I hear back!

File: src/js/letters/components/Address.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (raquelromano-gov) [reply src/js/letters/components/Address.jsx]: Oh, OK--@goldenmeanie and I had discussed that people with overseas military addresses are used to picking US as their country and looking for AA/AE/AP in a list of states. But I didn't catch that they also know to put APO/FPO/DPO in as the city type (I thought we'd be adding a conditional dropdown menu explicitly for the military post office code if a military state were chosen. So if that's what the research shows, that sounds fine to me.
